[libcxx] [test] Quote the python executable in the executor

mstorsjo opened the pull request:
This reapplies the change from
c218c80c730a14a1cbcebd588b18220a879702c6, which was lost in the refactoring in 78d649a417b48cb8a2ba2e755f0e7c8fb8b1bb83.

On Windows, the Python interpreter is by default installed in a path like "C:\Program Files\Python38\python.exe", which requires quoting when included in a concatenated command string (as opposed to a command list, where each argument is a separate element).

This doesn't show up in the CI environments as they use Python installed in a different directory, like C:\Python.

mstorsjo committed:
[libcxx] [test] Quote the python executable in the executor

This reapplies the change from
c218c80c730a14a1cbcebd588b18220a879702c6, which was lost in the
refactoring in 78d649a417b48cb8a2ba2e755f0e7c8fb8b1bb83.

On Windows, the Python interpreter is by default installed in
a path like "C:\Program Files\Python38\python.exe", which requires
quoting when included in a concatenated command string (as opposed
to a command list, where each argument is a separate element).

This doesn't show up in the CI environments as they use Python
installed in a different directory, like C:\Python.

arichardson reviewed (approved):
Should the run script path also be quoted?

arichardson commented on libcxx/utils/libcxx/test/params.py:
```diff
@@ -320,7 +320,7 @@ def getStdFlag(cfg, std):
         name="executor",
         type=str,
-        default=f"{sys.executable} {Path(__file__).resolve().parent.parent.parent / 'run.py'}",
+        default=f"\"{sys.executable}\" {Path(__file__).resolve().parent.parent.parent / 'run.py'}",
```
Could also use shutil.quote(sys.executable) to avoid unnecessary quoting?

mstorsjo replied:
Yes, that’d probably work too - I’ll try that out.

mstorsjo commented:
> Should the run script path also be quoted?

I guess it could; I guess most people would avoid checking out the source in such a path with spaces, but the default install location of Python usually is in a place with spaces. I’ll try adding that as well.

mstorsjo committed:
Switch to using a proper quote function

Don't use shlex.quote(), that's documented to only be relevant
for Unix (and it uses single quotes which aren't normally
tolerated by windows shells - even if the lit internal shell
does handle them). Additionally, shlex.quote() always quotes
paths that include backslashes, even if those don't need quoting
on Windows.

mstorsjo commented:
I updated the branch so that it uses a proper quote function, and quote both the script and the python executable. The proper python cross platform quoting for shell commands is `subprocess.list2cmdline`, somewhat non-obviously.

arichardson commented on libcxx/utils/libcxx/test/params.py:
```diff
@@ -320,7 +325,7 @@ def getStdFlag(cfg, std):
         name="executor",
         type=str,
-        default=f"\"{sys.executable}\" {Path(__file__).resolve().parent.parent.parent / 'run.py'}",
+        default=f"{quote(sys.executable)} {quote(str(Path(__file__).resolve().parent.parent.parent / 'run.py'))}",
```
Hmm actually it looks like list2cmdline is a private implementation detail: https://github.com/python/cpython/blob/6e97a9647ae028facb392d12fc24973503693bd6/Lib/subprocess.py#L66 so should probably use shlex.quote() instead.

mstorsjo replied:
Oh, good catch. I saw some discussions about whether this function should be made public (as they can't practically change it anyway, since lots of people seem to use it) - but for now I guess it'd be best to not use it.

If we're not using that one, I'm on the fence about whether we should go for `shlex.quote()`, always wrapping double quotes (like we did in cmake before), or maybe a custom function that wraps quotes around it if there are spaces in the string. `shlex.quote()` isn't generally a great fit, as it is documented to not be cross platform - it specifically does quoting for unix paths/shells. For Windows paths, it always quotes them as they contain backslashes. (Always quoting them isn't an issue in itself per se, though. And the fact that `shlex.quote()` uses single quotes instead of double quotes, is ok as long as we're using the lit internal shell instead of an actual Windows shell.)

mstorsjo replied:
Switched to `shlex.quote` now. Not ideal for Windows paths, but should work as long as we're using the lit internal shell.

ldionne reviewed (approved):
LGTM. It might be cleaner to use `shlex.quote` but either way makes sense to me.

mstorsjo committed:
Switch to using shlex.quote()

This isn't ideal on Windows, but subprocess.list2cmdline is
private.

As long as we're using the lit internal shell, the shlex quoting
should work fine.
